// @infra-blocks/zod-utils/lib/cjs/json/json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stringifiedJson = exports.stringifiedJsonSchema = exports.array = exports.object = exports.json = exports.primitive = void 0;
const zod_1 = require("zod");
const primitiveSchema = zod_1.z.union([
    zod_1.z.string(),
    zod_1.z.number(),
    zod_1.z.boolean(),
    zod_1.z.null(),
]);
function primitive() {
    return primitiveSchema;
}
exports.primitive = primitive;
const jsonSchema = zod_1.z.lazy(() => zod_1.z.union([primitive(), zod_1.z.array(jsonSchema), zod_1.z.record(jsonSchema)]));
function json() {
    return jsonSchema;
}
exports.json = json;
const objectSchema = zod_1.z.record(jsonSchema);
function object() {
    return objectSchema;
}
exports.object = object;
const arraySchema = zod_1.z.array(jsonSchema);
function array() {
    return arraySchema;
}
exports.array = array;
exports.stringifiedJsonSchema = zod_1.z
    .string()
    .transform((str, ctx) => {
    try {
        return JSON.parse(str);
    }
    catch (e) {
        ctx.addIssue({ code: "custom", message: "Invalid JSON" });
        return zod_1.z.NEVER;
    }
});
function stringifiedJson() {
    return exports.stringifiedJsonSchema;
}
exports.stringifiedJson = stringifiedJson;
